File: Search Algorithms/Java/BinarySearch.java
public class BinarySearch {
	public static void main(String args[]) {
		int a[] = {1, 5, 3, 4, 2};

		System.out.println("Searching for 3:");
		int i = binary_search(a, 3);
		System.out.println(i < 0 ? "Not found" : "Found at index" + i);

		System.out.println("Searching for 99:");
		binary_search(a, 99);
		System.out.println(i < 0 ? "Not found" : "Found at index" + i);
	}

	static void binary_search(int a[], k) {
		int min = 0, max = a.length;

		while (min < max) {
			int mid = (min + max) / 2;

			if (a[mid] < k) min = mid + 1;
			else if (a[mid] > k) min = mid - 1;
			else return mid;
		}
		return -1;
	}
}

File: Search Algorithms/Kotlin/BinarySearch.kt
fun <T : Comparable<T>> List<T?>.binarySearch(element: T?, fromIndex: Int = 0, toIndex: Int = size): Int {

    var begin = fromIndex
    var end = toIndex - 1

    while (begin <= end) {
        val mid = (begin + end) / 2
        val midVal = get(mid)
        val compare = compareValues(midVal, element)

        when {
            compare < 0 -> begin = mid + 1
            compare > 0 -> end = mid - 1
            else -> return mid
        }
    }
    return -1;
}



fun main(args: Array<String>) {
    val arr = intArrayOf(3, 2, 5, 9, 1).sorted()
    val testOne = 2
    val testTwo = 9

    val a = arr.binarySearch(testOne)
    println(if (a < 0) "$testOne not found" else "$testOne found at index $a")

    val b = arr.binarySearch(testTwo)
    println(if (b < 0) "$testTwo not found" else "$testTwo found at index $b")
}